crates/bevy_input/src/gamepad.rs

```diff
@@ -2,6 +2,27 @@ use crate::{Axis, Input};
 use bevy_ecs::event::{EventReader, EventWriter};
 use bevy_ecs::system::{Res, ResMut, Resource};
 use bevy_utils::{tracing::info, HashMap, HashSet};
+use thiserror::Error;
+
+/// Errors that occur when setting settings for gamepad input.
+#[derive(Error, Debug)]
+pub enum GamepadSettingsError {
+    #[error("the conditions -1.0 <= livezone_lowerbound <= deadzone_lowerbound <= 0.0 <= deadzone_upperbound <= livezone_upperbound <= 1.0 must hold true")]
+    InvalidAxisZoneSetting,
+    #[error("invalid threshold {0}, expected 0.0 <= threshold <= 2.0")]
+    InvalidAxisThresholdSetting(f32),
+    #[error("invalid release_threshold {0}, expected 0.0 <= release_threshold <= 1.0")]
+    InvalidButtonSettingReleaseThresholdOutOfRange(f32),
+    #[error("invalid press_threshold {0}, expected 0.0 <= press_threshold <= 1.0")]
+    InvalidButtonSettingPressThresholdOutOfRange(f32),
+    #[error("invalid parameter values release_threshold {} press_threshold {}, expected release_threshold <= press_threshold", .release_threshold, .press_threshold)]
+    InvalidButtonSettingReleaseThresholdGreaterThanPressThreshold {
+        press_threshold: f32,
+        release_threshold: f32,
+    },
+}
```

Comment:
> Could you add doc strings for each variant?

Each variant doesn't have a doc string?  The doc string for InvalidAxisZoneSetting is "the conditions -1.0 <= live....".  What am I missing?

> 
> I think the names could be made simpler, at least by not prefixing them by `Invalid`: the enum type already says it's an error. Maybe the word `Setting` can be not repeated too as it's also in the type name

Sound like good name changes.  Will implement.

Reply:
> Each variant doesn't have a doc string? The doc string for InvalidAxisZoneSetting is "the conditions -1.0 <= live....". What am I missing?

That's not the docstring, that's the parameter for the error derive from thiserror. It can't be reused as a docstring as it can be parameterised with the value of the error

Reply:
Added docstrings for enum variants.  Split `GamepadSettingsError` error into two enums, `ButtonSettingsError` and `AxisSettingsError`.